src/block_service/block_service_manager.cpp

```diff
@@ -84,133 +84,76 @@ block_filesystem *block_service_manager::get_block_filesystem(const std::string
-            return;
-        }
-        ddebug_f("download file({}) succeed, file_size = {}",
-                 local_file_name.c_str(),
-                 resp.downloaded_size);
-        download_err = ERR_OK;
-        download_file_size = resp.downloaded_size;
-    };
-
-    auto create_file_cb = [&local_dir,
-                           &download_err,
-                           &download_file_size,
-                           &download_file_callback_func,
-                           &tracker](const create_file_response &resp, const std::string &fname) {
-        if (resp.err != ERR_OK) {
-            derror_f("create file({}) failed with error({})", fname, resp.err.to_string());
-            download_err = resp.err;
-            return;
-        }
+    // local file exists
+    const std::string &local_file_name = utils::filesystem::path_combine(local_dir, file_name);
+    if (utils::filesystem::file_exists(local_file_name)) {
+        ddebug_f("local file({}) has been downloaded", local_file_name);
+        return ERR_OK;
+    }
 
-        block_file *bf = resp.file_handle.get();
-        if (bf->get_md5sum().empty()) {
-            derror_f("file({}) doesn't exist on remote file provider", bf->file_name());
-            download_err = ERR_CORRUPTION;
-            return;
-        }
+    task_tracker tracker;
 
-        const std::string &local_file_name = utils::filesystem::path_combine(local_dir, fname);
-        // local file exists
-        if (utils::filesystem::file_exists(local_file_name)) {
-            std::string current_md5;
-            error_code e = utils::filesystem::md5sum(local_file_name, current_md5);
-            if (e != ERR_OK || current_md5 != bf->get_md5sum()) {
-                if (e != ERR_OK) {
-                    dwarn_f("calculate file({}) md5 failed, should remove and redownload it",
-                            local_file_name);
-                } else {
-                    dwarn_f("local file({}) is different from remote file({}), md5: local({}) VS "
-                            "remote({}), should remove and redownload it",
-                            local_file_name,
-                            bf->file_name(),
-                            current_md5,
-                            bf->get_md5sum());
-                }
-                if (!utils::filesystem::remove_path(local_file_name)) {
-                    derror_f("failed to remove file({})", local_file_name);
-                    download_err = e;
-                    return;
-                }
-            } else {
-                download_err = ERR_OK;
-                download_file_size = bf->get_size();
-                ddebug_f("local file({}) has been downloaded, file size = {}",
-                         local_file_name,
-                         download_file_size);
-                return;
-            }
+    // Create a block_file object.
+    const std::string remote_file_name = utils::filesystem::path_combine(remote_dir, file_name);
```

Comment:
const std::string &remote_file_name

Reply:
remote_file_name is a new value, not a reference to an existing value.

Reply:
so why did you use `  const std::string &local_file_name` in L121?